src/audio/module_adapter/module_adapter_ipc4.c

```diff
@@ -38,19 +38,39 @@ int module_adapter_init_data(struct comp_dev *dev,
 			     const struct comp_ipc_config *config,
 			     const void *spec)
 {
-	if (dev->drv->type == SOF_COMP_MODULE_ADAPTER) {
-		const struct ipc_config_process *ipc_module_adapter = spec;
+	const struct ipc_config_process *cp = spec;
+	const struct ipc4_base_module_extended_cfg *cfg = (void *)cp->data;
+	size_t cfgsz = cp->size;
 
-		dst->init_data = ipc_module_adapter->data;
-		dst->size = ipc_module_adapter->size;
-		dst->avail = true;
+	assert(dev->drv->type == SOF_COMP_MODULE_ADAPTER);
+	if (cp->size < sizeof(cfg->base_cfg))
+		return -EINVAL;
 
-		memcpy_s(&dst->base_cfg,  sizeof(dst->base_cfg), ipc_module_adapter->data,
-			 sizeof(dst->base_cfg));
-	} else {
-		dst->init_data = spec;
+	dst->base_cfg = cfg->base_cfg;
+	dst->size = cp->size;
+
+	if (cp->size >= sizeof(*cfg)) {
```

Comment:
Well to be fair, this check is needed to prevent a buffer overrun during parsing.  It's true that there's also a typesafety issue in knowing "what kind of extra data do we have", but that's a protocol bug not really addressed by this PR :)